interfaces/tests/file_int/file_int.c

```diff
@@ -0,0 +1,427 @@
+#include <cstdlib>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#else
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#endif
+
+
+#include "testrunnerswitcher.h"
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+#include "azure_c_pal/sync.h"
+#include "azure_c_pal/file.h"
+#include "azure_c_pal/interlocked.h"
+
+
+TEST_DEFINE_ENUM_TYPE(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_RESULT)
+TEST_DEFINE_ENUM_TYPE(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_RESULT)
+
+typedef struct WRITE_COMPLETE_CONTEXT_TAG
+{
+    int32_t pre_callback_value;
+    volatile int32_t value;
+    int32_t post_callback_value;
+    bool did_write_succeed;
+}WRITE_COMPLETE_CONTEXT;
+
+typedef struct READ_COMPLETE_CONTEXT_TAG
+{
+    int32_t pre_callback_value;
+    volatile int32_t value;
+    int32_t post_callback_value;
+    bool did_read_succeed;
+}READ_COMPLETE_CONTEXT;
+
+static void write_callback(void* context, bool is_successful) {
+    WRITE_COMPLETE_CONTEXT* write_context = (WRITE_COMPLETE_CONTEXT*)context;
+    ASSERT_ARE_EQUAL(int32_t, write_context->pre_callback_value, write_context->value);
+    interlocked_exchange(&(write_context->value), write_context->post_callback_value);
+    write_context->did_write_succeed = is_successful;
+    wake_by_address_single(&write_context->value);
+}
+
+static void read_callback(void* context, bool is_successful) {
+    READ_COMPLETE_CONTEXT* read_context = (READ_COMPLETE_CONTEXT*)context;
+    ASSERT_ARE_EQUAL(int32_t, read_context->pre_callback_value, read_context->value);
+    interlocked_exchange(&(read_context->value), read_context->post_callback_value);
+    read_context->did_read_succeed = is_successful;
+    wake_by_address_single(&(read_context->value));
+}
+
+
+BEGIN_TEST_SUITE(file_int)
+
+
+TEST_SUITE_INITIALIZE(a)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+}
+
+TEST_SUITE_CLEANUP(b)
+{
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(c)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+}
+
+TEST_FUNCTION_CLEANUP(d)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Tests_SRS_FILE_43_003: [If a file with name full_file_name does not exist, file_create shall create a file with that name.]*/
+/*Tests_SRS_FILE_43_001 : [file_create shall open the file named full_file_name for asynchronous operationsand return its handle.]*/
+/*Tests_SRS_FILE_43_006 : [file_destroy shall wait for all pending I / O operations to complete.]*/
+/*Tests_SRS_FILE_43_007 : [file_destroy shall close the file handle handle.]*/
+/*Tests_SRS_FILE_43_014 : [file_write_async shall enqueue a write request to write source's content to the position offset in the file. ]*/
+/*Tests_SRS_FILE_43_041:[If position + size is greater than the size of the file and the call to write is succesfull, file_write_async shall grow the file to accomodate the write.]*/

```

Comment:
>hall grow the file [](start = 136, length = 18)

Can't really test this unless you can delete the file if it exists as part of the setup